accepted/2024/trusted-publishers-oidc-for-nuget-push.technical.md

```diff
@@ -0,0 +1,189 @@
+token](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect#understanding-the-oidc-token)
+document.
+
+Other Trusted Publishers like Azure DevOps or Bitbucket should have sufficient token claims so both the registry
+(NuGet.org) and the package author are certain that only proper workload identity tokens are traded for privileged
+short-lived API keys.
+
+## Data relationships (persisting trust policies, schema changes)
+
+A new SQL table will be added to the NuGetGallery database to store trust policies. The table should be generic enough
+to allow us to add additional Trusted Publishers without a DB schema change (ideally). The trust policy DB record with
+have a foreign key to the `Users` table (containing both user and organization records) but will be restricted from
+associating with organization records by the application. Many of the columns will be shared with the `Credentials`
+table in order to express the scoping rules to be copied into the short-lived API key.
+
+A nullable column will be added to the `Credentials` table to refer to the trust policy used to create the short-lived
+API key. This will allow package publish operations to audit their related trust policy information.
+
+Deleting a trust policy should have the effect of deleting all related short-lived API keys.
+
+## Trading an OIDC token for an API key
+
+A new endpoint will be needed for trading an bearer token (OIDC token, a JWT) for an API key. The endpoint URL will be
+discoverable via the [V3 service index](https://learn.microsoft.com/en-us/nuget/api/overview#service-index) and a new
+resource type which is `ApiKeyService/1.0.0`. For NuGet.org, the service index is available at
+`https://api.nuget.org/v3/index.json`. The new resource URL be something like `https://www.nuget.org/api/v2/api-key`.
+
+```
+POST /api/v2/api-key HTTP/1.1
+Host: www.nuget.org
+Authorization: Bearer {OIDC token}
+Content-Type: application/json
+
+{
+   "username": "{username of user with trust policy}"
+}
+```
+
+The response will look like this:
+
+```
+HTTP/1.1 200 OK
+Content-Type: application/json
+
+{
+   "api_key": "{short lived API key in clear text}",
+   "expires": "{ISO 8601 timestamp of expiration}"
+}
+```
+
+Authorization failures on this endpoint must return HTTP 401 Unauthorized with an `WWW-Authenticate: Bearer` response
+header.
+
+The package source MUST NOT opt to return the an existing compatible API key (i.e. it must not cache the API key for
+subsequent calls). To do so would require the original API key to be stored in plain text. NuGet.org API keys are hashed
+prior to storage (much like standard recommendations around storing passwords). The the package source has concerns on
+scalability it must opt to rate limit the endpoint instead of caching. NuGet.org will rate limit the endpoint to 1 API
+key created per 30 seconds, per user.
```

Comment:
I don't grok how this limit aligns with someone publishing many packages in a short period of time.  Consider the .NET team publishing packages in bulk to NuGet.org.  Will that translate to hundreds or thousands of API calls within a short timeframe?  Or, is the API key cached and reused on the client side?

It would be good to discuss here how you see this limitation not being a problem for various publishing scenarios.

Reply:
> Will that translate to hundreds or thousands of API calls within a short timeframe? Or, is the API key cached and reused on the client side?

I expect the API key to be cached on the client side. In most workflows (no custom scripting), this will be done by a single `nuget/login` workflow step with a single output value, which is the short lived API key. This could be used one or more push operations (all referring to the single step output == single API key). I imagine most users will do `dotnet nuget push *.nupkg` which internally does multiple push operations with a single API key provided via CLI arg.

Reply:
Suppose `dotnet nuget push *.nupkg` takes longer than 15 minutes.  Imagine lots of packages, slow network, whatever.  Will NuGet client automatically re-request a new token via `nuget/login` midstream, as necessary?

Reply:
Not with this version of the proposal. See the https://github.com/NuGet/Home/blob/jver-oidc/accepted/2024/trusted-publishers-oidc-for-nuget-push.md#enable-nuget-authentication-provider-integration-to-avoid-token-expiration-for-long-workflows.

The user will be able to work around the issue with multiple `nuget/login` GitHub Actions steps and breaking up the push steps. It's not a great solution, but the data suggests we can pick a single short lifetime (15 minutes) and cover the vast majority of cases.